sphinx/source/docs/user_guide/data.rst

```diff
@@ -101,20 +127,49 @@ the ``source`` argument:
     p = figure()
     p.circle(x='x_values', y='y_values', source=source)
 
-To use a ColumnDataSource with a plotting function, you need to pass at least
-these three arguments:
+Modifying a ColumnDataSource
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ``x``: the name of the ColumnDataSource's column that contains the data for
-  the x values of your plot
-* ``y``: the name of the ColumnDataSource's column that contains the data for
-  the y values of your plot
-* ``source``: the name of the ColumnDataSource that contains the columns you
-  just referenced for the ``x`` and ``y`` arguments.
+To modify the data of an existing ColumnDataSource, update the ``.data``
+property of your ColumnDataSource object:
+
+* To add a new column to an existing ColumnDataSource:
+
+  .. code-block:: python
+
+    new_sequence = [8, 1, 4, 7, 3]
+    source.data["new_column"] = new_sequence
+
+  The length of the column you are adding must match the length of the existing
+  columns.
+
+* To replace data in an existing column within your ColumnDataSource, assign
+  your new value to the respective index position:
+
+  .. code-block:: python
+
+    source.data["new_column"][2] = 5
```

Comment:
I think we should remove this middle option.  This kind of "in place" update is best handled by using `patch`

cc @bokeh/core for any other opinions though

Reply:
I have removed the middle option for now.